gcc/config/riscv/bsg_vanilla.md

```diff
@@ -22,76 +22,82 @@
 
 (define_automaton "bsg_vanilla")
 (define_cpu_unit "bsg_vanilla_alu" "bsg_vanilla")
-(define_cpu_unit "bsg_vanilla_imuldiv" "bsg_vanilla")
-(define_cpu_unit "bsg_vanilla_fdivsqrt" "bsg_vanilla")
 
+;; ALU
 (define_insn_reservation "bsg_vanilla_alu" 1
   (and (eq_attr "tune" "bsg_vanilla")
        (eq_attr "type" "unknown,const,arith,shift,slt,multi,auipc,nop,logical,move"))
   "bsg_vanilla_alu")
 
+;; remote load
 (define_insn_reservation "bsg_vanilla_remote_load" 32
   (and (eq_attr "tune" "bsg_vanilla")
        (eq_attr "type" "load,fpload")
        (eq_attr "remote_mem_op" "yes"))
   "bsg_vanilla_alu")
 
-(define_insn_reservation "bsg_vanilla_load" 3
+;; integer local load
+(define "bsg_vanilla_load" 2
   (and (eq_attr "tune" "bsg_vanilla")
-       (eq_attr "type" "load,fpload")
+       (eq_attr "type" "load")
        (eq_attr "remote_mem_op" "no"))
   "bsg_vanilla_alu")
 
-(define_insn_reservation "bsg_vanilla_store" 1
+;; FP local load
+(define "bsg_vanilla_fpload" 3
   (and (eq_attr "tune" "bsg_vanilla")
-       (eq_attr "type" "store,fpstore"))
+       (eq_attr "type" "fpload")
+       (eq_attr "remote_mem_op" "no"))
   "bsg_vanilla_alu")
 
-(define_insn_reservation "bsg_vanilla_xfer" 3
+;; store
+(define_insn_reservation "bsg_vanilla_store" 1
   (and (eq_attr "tune" "bsg_vanilla")
-       (eq_attr "type" "mfc,mtc,fcvt,fmove,fcmp"))
+       (eq_attr "type" "store,fpstore"))
   "bsg_vanilla_alu")
 
+;; branch
 (define_insn_reservation "bsg_vanilla_branch" 1
   (and (eq_attr "tune" "bsg_vanilla")
        (eq_attr "type" "branch,jump,call"))
   "bsg_vanilla_alu")
 
-(define_insn_reservation "bsg_vanilla_imul" 10
+;; imul
+(define_insn_reservation "bsg_vanilla_imuldiv" 33
   (and (eq_attr "tune" "bsg_vanilla")
-       (eq_attr "type" "imul"))
-  "bsg_vanilla_imuldiv*10")
+       (eq_attr "type" "imul,idiv"))
+  "bsg_vanilla_alu*33")
 
-(define_insn_reservation "bsg_vanilla_idivsi" 34
+;; i2f 
+(define_insn_reservation "bsg_vanilla_i2f" 5
   (and (eq_attr "tune" "bsg_vanilla")
-       (and (eq_attr "type" "idiv")
-	    (eq_attr "mode" "SI")))
-  "bsg_vanilla_imuldiv*34")
-
-(define_insn_reservation "bsg_vanilla_idivdi" 66
-  (and (eq_attr "tune" "bsg_vanilla")
-       (and (eq_attr "type" "idiv")
-	    (eq_attr "mode" "DI")))
-  "bsg_vanilla_imuldiv*66")
+       (eq_attr "type" "mtc"))
+  "bsg_vanilla_alu")
 
-(define_insn_reservation "bsg_vanilla_fmul_single" 5
+;; f2i
+(define_insn_reservation "bsg_vanilla_f2i" 1
   (and (eq_attr "tune" "bsg_vanilla")
-       (and (eq_attr "type" "fadd,fmul,fmadd")
-	    (eq_attr "mode" "SF")))
+       (eq_attr "type" "mfc"))
   "bsg_vanilla_alu")
 
-(define_insn_reservation "bsg_vanilla_fmul_double" 7
+;; fpu_float
+;; note: WAW hazards for FPU not currently modeled.
+;; mostly it applies after branch.
+(define_insn_reservation "bsg_vanilla_fpu_float" 5
```

Comment:
maybe a comment that explains why 5? Isn't it 3?

Reply:
For the old pipeline, FP_EXE -> F -> P -> U -> FP_WB. It's total of 5 cycles.